File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/utils/SafeIntent.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at http://mozilla.org/MPL/2.0/.
 */

/*
 * Extracted from Fennec/Firefox:
 * mobile/android/geckoview/src/main/java/org/mozilla/gecko/mozglue/SafeIntent.java
 */
package org.mozilla.focus.utils;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.util.Log;

import java.util.ArrayList;
import java.util.Set;

/**
 * External applications can pass values into Intents that can cause us to crash: in defense,
 * we wrap {@link Intent} and catch the exceptions they may force us to throw. See bug 1090385
 * for more.
 */
public class SafeIntent {
    private static final String LOGTAG = "Gecko" + SafeIntent.class.getSimpleName();

    private final Intent intent;

    public SafeIntent(final Intent intent) {
        this.intent = intent;
    }

    public boolean hasExtra(String name) {
        try {
            return intent.hasExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't determine if intent had an extra: OOM. Malformed?");
            return false;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't determine if intent had an extra.", e);
            return false;
        }
    }

    public boolean getBooleanExtra(final String name, final boolean defaultValue) {
        try {
            return intent.getBooleanExtra(name, defaultValue);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return defaultValue;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return defaultValue;
        }
    }

    public int getIntExtra(final String name, final int defaultValue) {
        try {
            return intent.getIntExtra(name, defaultValue);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return defaultValue;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return defaultValue;
        }
    }

    public String getStringExtra(final String name) {
        try {
            return intent.getStringExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    @Nullable
    public Bundle getExtras() {
        try {
            return intent.getExtras();
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    public SafeBundle getBundleExtra(final String name) {
        try {
            final Bundle bundle = intent.getBundleExtra(name);
            if (bundle != null) {
                return new SafeBundle(bundle);
            } else {
                return null;
            }
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    public CharSequence getCharSequenceExtra(final String name) {
        try {
            return intent.getCharSequenceExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    public <T extends Parcelable> T getParcelableExtra(final String name) {
        try {
            return intent.getParcelableExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    public <T extends Parcelable> ArrayList<T> getParcelableArrayListExtra(final String name) {
        try {
            return intent.getParcelableArrayListExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e);
            return null;
        }
    }

    public String getAction() {
        return intent.getAction();
    }

    public int getFlags() {
        return intent.getFlags();
    }

    public String getDataString() {
        try {
            return intent.getDataString();
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent data string: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent data string.", e);
            return null;
        }
    }

    public ArrayList<String> getStringArrayListExtra(final String name) {
        try {
            return intent.getStringArrayListExtra(name);
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent data string: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent data string.", e);
            return null;
        }
    }

    public Uri getData() {
        try {
            return intent.getData();
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent data: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent data.", e);
            return null;
        }
    }

    public Set<String> getCategories() {
        try {
            return intent.getCategories();
        } catch (OutOfMemoryError e) {
            Log.w(LOGTAG, "Couldn't get intent categories: OOM. Malformed?");
            return null;
        } catch (RuntimeException e) {
            Log.w(LOGTAG, "Couldn't get intent categories.", e);
            return null;
        }
    }

    public boolean isLauncherIntent() {
        final Set<String> intentCategories = intent.getCategories();
        return (intentCategories != null && intentCategories.contains(Intent.CATEGORY_LAUNCHER) && Intent.ACTION_MAIN.equals(intent.getAction()));
    }

    public Intent getUnsafe() {
        return intent;
    }
}

File: mobile/android/android-components/components/support/utils/src/main/java/mozilla/components/utils/SafeIntent.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.components.utils

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcelable
import android.util.Log

import java.util.ArrayList

/**
 * External applications can pass values into Intents that can cause us to crash: in defense,
 * we wrap [Intent] and catch the exceptions they may force us to throw. See bug 1090385
 * for more.
 */
class SafeIntent(val unsafe: Intent) {

    val extras: Bundle?
        get() {
            try {
                return unsafe.extras
            } catch (e: OutOfMemoryError) {
                Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
                return null
            } catch (e: RuntimeException) {
                Log.w(LOGTAG, "Couldn't get intent extras.", e)
                return null
            }
        }

    val action: String?
        get() = unsafe.action

    val flags: Int
        get() = unsafe.flags

    val dataString: String?
        get() {
            try {
                return unsafe.dataString
            } catch (e: OutOfMemoryError) {
                Log.w(LOGTAG, "Couldn't get intent data string: OOM. Malformed?")
                return null
            } catch (e: RuntimeException) {
                Log.w(LOGTAG, "Couldn't get intent data string.", e)
                return null
            }
        }

    val data: Uri?
        get() {
            try {
                return unsafe.data
            } catch (e: OutOfMemoryError) {
                Log.w(LOGTAG, "Couldn't get intent data: OOM. Malformed?")
                return null
            } catch (e: RuntimeException) {
                Log.w(LOGTAG, "Couldn't get intent data.", e)
                return null
            }
        }

    val categories: Set<String>?
        get() {
            try {
                return unsafe.categories
            } catch (e: OutOfMemoryError) {
                Log.w(LOGTAG, "Couldn't get intent categories: OOM. Malformed?")
                return null
            } catch (e: RuntimeException) {
                Log.w(LOGTAG, "Couldn't get intent categories.", e)
                return null
            }
        }

    val isLauncherIntent: Boolean
        get() {
            val intentCategories = unsafe.categories
            return intentCategories != null && intentCategories.contains(Intent.CATEGORY_LAUNCHER) && Intent.ACTION_MAIN == unsafe.action
        }

    fun hasExtra(name: String): Boolean {
        try {
            return unsafe.hasExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't determine if intent had an extra: OOM. Malformed?")
            return false
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't determine if intent had an extra.", e)
            return false
        }
    }

    fun getBooleanExtra(name: String, defaultValue: Boolean): Boolean {
        try {
            return unsafe.getBooleanExtra(name, defaultValue)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return defaultValue
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return defaultValue
        }
    }

    fun getIntExtra(name: String, defaultValue: Int): Int {
        try {
            return unsafe.getIntExtra(name, defaultValue)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return defaultValue
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return defaultValue
        }
    }

    fun getStringExtra(name: String): String? {
        try {
            return unsafe.getStringExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return null
        }
    }

    fun getBundleExtra(name: String): SafeBundle? {
        try {
            val bundle = unsafe.getBundleExtra(name)
            return if (bundle != null) {
                SafeBundle(bundle)
            } else {
                null
            }
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return null
        }
    }

    fun getCharSequenceExtra(name: String): CharSequence? {
        try {
            return unsafe.getCharSequenceExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return null
        }
    }

    fun <T : Parcelable> getParcelableExtra(name: String): T? {
        try {
            return unsafe.getParcelableExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return null
        }
    }

    fun <T : Parcelable> getParcelableArrayListExtra(name: String): ArrayList<T>? {
        try {
            return unsafe.getParcelableArrayListExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent extras: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent extras.", e)
            return null
        }
    }

    fun getStringArrayListExtra(name: String): ArrayList<String>? {
        try {
            return unsafe.getStringArrayListExtra(name)
        } catch (e: OutOfMemoryError) {
            Log.w(LOGTAG, "Couldn't get intent data string: OOM. Malformed?")
            return null
        } catch (e: RuntimeException) {
            Log.w(LOGTAG, "Couldn't get intent data string.", e)
            return null
        }
    }

    companion object {
        private val LOGTAG = "Gecko" + SafeIntent::class.java.simpleName
    }
}
